Tests/scripts/collect_tests/utils.py

```diff
@@ -0,0 +1,275 @@
+from packaging._structures import NegativeInfinityType, InfinityType
+
+from exceptions import (DeprecatedPackException, BlankPackNameException,
+                        NonDictException, NonexistentPackException,
+                        NoTestsConfiguredException, NotUnderPackException,
+                        SkippedPackException, UnsupportedPackException)
+from logger import logger
+from packaging import version
+from packaging.version import Version
+from path_manager import PathManager
+
+
+def find_pack_folder(path: Path) -> Path:
+    """
+    >>> find_pack_folder(Path('root/Packs/MyPack/Integrations/MyIntegration/MyIntegration.yml'))
+    PosixPath('root/Packs/MyPack')
+    >>> find_pack_folder(Path('Packs/MyPack1/Scripts/MyScript/MyScript.py')).name
+    'MyPack1'
+    >>> find_pack_folder(Path('Packs/MyPack2/Scripts/MyScript')).name
+    'MyPack2'
+    >>> find_pack_folder(Path('Packs/MyPack3/Scripts')).name
+    'MyPack3'
+    >>> find_pack_folder(Path('Packs/MyPack4')).name
+    'MyPack4'
+    """
+    if 'Packs' not in path.parts:
+        raise NotUnderPackException(path)
+    if path.parent.name == 'Packs':
+        return path
+    return path.parents[len(path.parts) - (path.parts.index('Packs')) - 3]
+
+
+@dataclass
+class VersionRange:
+    min_version: Version | NegativeInfinityType
+    max_version: Version | InfinityType
+
+    def __contains__(self, item):
+        return self.min_version <= item <= self.max_version
+
+    def __repr__(self):
+        return f'{self.min_version} -> {self.max_version}'
+
+    def __or__(self, other: Optional['VersionRange']) -> 'VersionRange':
+        if other is None or other.is_default or self.is_default:
+            return self
+
+        self.min_version = min(self.min_version, other.min_version)
+        self.max_version = max(self.max_version, other.max_version)
+
+        return self
+
+    @property
+    def is_default(self):
+        return self.min_version == version.NegativeInfinity and self.max_version == version.Infinity
+
+
+class Machine(Enum):
+    V6_5 = Version('6.5')
+    V6_6 = Version('6.6')
+    V6_8 = Version('6.8')
+    MASTER = 'master'
+    NIGHTLY = 'nightly'
+
+    @staticmethod
+    def numeric_machines() -> tuple['Machine', ...]:
+        return tuple(machine for machine in Machine if isinstance(machine.value, Version))
+
+    @staticmethod
+    def get_suitable_machines(version_range: Optional[VersionRange], run_nightly: bool) -> tuple['Machine', ...]:
+        result: list[Machine] = [Machine.MASTER]
+
+        if not version_range:
+            version_range = VersionRange(version.NegativeInfinity, version.Infinity)
+
+        result.extend(machine for machine in Machine.numeric_machines() if machine.value in version_range)
+
+        if run_nightly:
+            result.append(Machine.NIGHTLY)
+
+        return tuple(result)
+
+    def __repr__(self):
+        return f'Server {self.value}'
+
+
+class DictBased:
```

Comment:
general comment, please add short doc string to the class in in this file that explain for each one of them what is their purpose